osquery/events/pathset.h

```diff
@@ -11,149 +11,142 @@
 #include <osquery/filesystem.h>
 
 namespace osquery {
 
 /**
- * @brief multiset based implementation for path search.
+ * @brief Vector based implemention for path search.
  *
- * 'multiset' is used because with patterns we can serach for equivalent keys.
- * Since  '/This/Path/is' ~= '/This/Path/%' ~= '/This/Path/%%' (equivalent).
+ * PathSet applies the following policy -
+ * patternedPath -    Path can contain pattern '%' and '%%'.
+ *                    Path components containing partial patterns are also
+ *                    supported e.g. '/This/Path/xyz%' or '/This/Path/%xyz'
+ *                    or '/This/Path/%xyz%' but '/This/Path/xy%z'
+ *                    is not supported i.e. 'xy%z' is considered as normal
+ *                    string.
  *
- * multiset is protected by lock. It is threadsafe.
- *
- * PathSet can take any of the two policies -
- * 1. patternedPath - Path can contain pattern '%' and '%%'.
- *                    Path components containing only '%' and '%%' are supported
- *                    e.g. '/This/Path/%'.
- *                    Path components containing partial patterns are not
- *                    supported e.g. '/This/Path/xyz%' ('xyz%' will not be
- *                    treated as pattern).
  */
 template <typename PathType>
 class PathSet : private boost::noncopyable {
  public:
   void insert(const std::string& str) {
     auto pattern = str;
     replaceGlobWildcards(pattern);
-    auto vpath = PathType::createVPath(pattern);
+    auto path = PathType::createPath(std::move(pattern));
 
     WriteLock lock(mset_lock_);
-    for (auto& path : vpath) {
-      paths_.insert(std::move(path));
-    }
+    patterns_.push_back(std::move(path));
   }
 
   bool find(const std::string& str) const {
     auto path = PathType::createPath(str);
 
     ReadLock lock(mset_lock_);
-    if (paths_.find(path) != paths_.end()) {
-      return true;
+    for (const auto& pattern : patterns_) {
+      if (compare(pattern, path)) {
+        return true;
+      }
     }
     return false;
   }
 
   void clear() {
     WriteLock lock(mset_lock_);
-    paths_.clear();
+    patterns_.clear();
   }
 
   bool empty() const {
     ReadLock lock(mset_lock_);
-    return paths_.empty();
+    return patterns_.empty();
   }
 
  private:
   typedef typename PathType::Path Path;
   typedef typename PathType::Compare Compare;
-  std::multiset<Path, Compare> paths_;
+  std::vector<Path> patterns_;
   mutable Mutex mset_lock_;
+  Compare compare;
 };
 
 class patternedPath {
  public:
   typedef boost::tokenizer<boost::char_separator<char>> tokenizer;
   typedef std::vector<std::string> Path;
-  typedef std::vector<Path> VPath;
   struct Compare {
+    bool compareStrings(boost::string_view pattern,
+                        boost::string_view str) const {
+      if (pattern[0] == '*' && pattern[pattern.size() - 1] == '*') {
+        pattern = pattern.substr(1, pattern.size() - 2);
+        if (str.size() >= pattern.size() &&
```

Comment:
str.size() >= pattern.size()  not needed

Reply:
*abcdef*
abcde

in the above example I have to make sure str.size() >= pattern.size() before comparison else there is not point for going for comparison since "abcdef" != "abcde"